### PR TITLE
Fix Fetcher's message skipping 

### PR DESCRIPTION
**Pull request by nguyenquangminh0711**
In the current Fetcher and Consumer, the current implementation has a bug that leads to the situation where a huge amount of messages are skipped, without being processed. The root cause is that the fetcher doesn't handle the case where a fetched batch is empty. 

In normal scenario, it updates the current offset based on the fetched batch last offset. When the batch is empty, it updates the current offset to the highwater mark offset, which is the end of the queue. The empty batch didn't exist before. However, newer versions of Kafka (after 0.11.0, I guess) do return the empty batches in some cases:
- Control batches of transactional message production
- Throttled fetch request
- Incomplete message sets which exceed the max byte limit
- etc.
My fix is just not to update current offset when the batch is empty. In case there is nothing more to fetch, the consumers just don't update the offset, instead of using the highwater mark offset like current.

Fix https://github.com/zendesk/ruby-kafka/issues/624
Merge merge before https://github.com/zendesk/ruby-kafka/pull/608

Reference: two cases that return empty sets. There would be more cases.
screen shot 2018-08-07 at 14 24 59

### CODE REVIEW

**Comment by Elyahou** [comment]
@nguyenquangminh0711 Can this issue be related to https://github.com/zendesk/ruby-kafka/issues/555 ?

**Comment by mensfeld** [comment]
@nguyenquangminh0711 @dasch can we bcakport it to `0.6`?

File: lib/kafka/client.rb [hunk 1]
```diff
@@ -477,7 +477,9 @@ def each_message(topic:, start_from_beginning: true, max_wait_time: 5, min_bytes
 
         batches.each do |batch|
           batch.messages.each(&block)
-          offsets[batch.partition] = batch.last_offset + 1
+          unless batch.messages.empty?
```
Comment (dasch) [inline lib/kafka/client.rb]: You should be able to do `batch.empty?`.

**Comment by nguyenquangminh0711** [comment]
@dasch Updated 💪

**Comment by dasch** [comment]
:heart:

**Comment by mainameiz** [comment]
How about to add specs to test such edge cases, guys?
